util/coverage/common.py

```diff
@@ -0,0 +1,279 @@
+                             check=True)
+    except subprocess.CalledProcessError as e:
+        log.error(f"stdout: {e.stdout if e.stdout else '(empty)'}")
+        log.error(f"stderr: {e.stderr if e.stderr else '(empty)'}")
+        raise
+    log.debug(f"stdout: {res.stdout if res.stdout else '(empty)'}")
+    log.debug(f"stderr: {res.stderr if res.stderr else '(empty)'}")
+    return [line for line in res.stdout.splitlines() if line]
+
+
+def create_out_dir(out_root_dir: Path, out_sub_dir: PurePath) -> Path:
+    """Create the directory for coverage artifacts.
+
+    Coverage artifacts will be saved in `out_root_dir`/<HEAD_TIMESTAMP>-<HEAD_HASH>/`out_sub_dir`/.
+
+    Args:
+        out_root_dir: Root of the output directory.
+        out_sub_dir: Directory to create under <HEAD_TIMESTAMP>-<HEAD_HASH>.
+
+    Returns:
+        Path where to save the coverage artifacts.
+    """
+    [head_hash] = run("git", "rev-parse", "HEAD")
+    [head_timestamp] = run("git", "show", "-s", "--format=%ct", "HEAD")
+    # Put timestamp first to be able to sort chronologically.
+    out_dir = out_root_dir / f"{head_timestamp}-{head_hash}" / out_sub_dir
+    out_dir.mkdir(parents=True, exist_ok=True)
+    return out_dir
+
+
+def instrument_device_libs(device_libs: List[str], config: str) -> List[str]:
+    """Instrument device libraries.
+
+    A coverage report can be created from multiple object files as long as they have
+    coverage data. However, unused inline functions can trigger confusing hash-mismatch
+    warnings and functions that appear in more than one object file can cause
+    undercounting. This function builds the given device libraries with coverage
+    instrumentation, gathers and returns their object files which can then be combined
+    into a single library to avoid such problems.
+
+    Args:
+        device_libs: List of device libraries to be instrumented.
+        config: Bazel configuration to use.
+
+    Returns:
+        Instrumented object files of the given device libraries.
+    """
+    log.info(f"libraries to be instrumented: {pformat(device_libs)}")
+    run(BAZEL, "build", f"--config={config}", *device_libs)
+    obj_files = run(
+        BAZEL, "cquery", f"--config={config}", f"({' + '.join(device_libs)})",
+        "--output=starlark",
+        ("--starlark:expr='\\n'.join("
+         "[f.path for f in target.output_groups.compilation_outputs.to_list()]"
+         ")"))
+    log.info(f"object files with coverage data: {pformat(obj_files)}")
+    return obj_files
+
+
+def get_test_log_dirs(test_targets: List[str]) -> List[Path]:
+    """Get log direactories for the given test targets.
+
+    Args:
+        test_targets: Test targets.
+
+    Returns:
+        Log directories of the the given test targets.
+    """
+    [test_log_dir] = run(
+        BAZEL,
+        "info",
+        "bazel-testlogs",
+    )
+    return [Path(test_log_dir) / t[2:].replace(':', '/') for t in test_targets]
```

Comment:
Nit: Can this be an explicit loop? Why do we drop the first two elements of `t`? Maybe that deserves a comment.